### PR TITLE
Improvements to EffAssert Logging

### PR DESCRIPTION
**Pull request by sovdeeth**
### Description

Adds the ability to provide expected and received values for an assertion that will be provided in the error message.

```
assert <.+> [(1:to fail)] with [error] %string%, expected [value] %object%, [and] (received|got) [value] %object%
```

For assertions using conditions that implement VerboseAssert (and that don't provide custom expected/got values), EffAssert will automatically grab expected/got messages from the condition itself. This will not work reliably for expressions that change between calls, like random expressions or ExprNow, as the assert effect calls getSingle/getAll a second time. For those scenarios, using a variable or a custom expected/got message is recommended.

So far, only CondCompare and CondIsSet implement VerboseAssert, but any condition that has significant use in tests is a valid target for implementation.

Suggestions are extremely welcome for a better interface name. This is the best kenzie and I could come up with in 5 minutes.

Example:
```applescript
assert 1 is 2 with "failed number comparison" # sends "failed number comparison (Expected a value equal to 2, got 1)."
assert 1 is 2 or 3 with "failed number list comparison" # sends "failed number list comparison (Expected a value equal to 2 or 3, got 1)."

assert name of player is "tony" with "failed name check" # sends "failed name check (Expected a value equal to tony, got Sahvde)."

assert isNaN(difference between NaN value and NaN value) is true with "difference between NaN and NaN is not NaN", expected NaN value, got difference between NaN value and NaN value
# sends "difference between NaN and NaN is not NaN (Expected NaN value, got <number>)."
```

This PR also adds line numbers to the test error message:
```patch
- [ScriptName.sk]
+ [ScriptName.sk, line 24]
```

### Todo:

- [x] add expected/got value
- [x] add line numbers
- [x] clean up changes

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

**Review by Moderocky — APPROVED** [review]
Very welcome addition.

File: src/main/java/ch/njol/skript/test/runner/EffAssert.java [hunk 1]
```diff
@@ -75,8 +96,8 @@ public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelaye
 		} finally {
 			logHandler.stop();
 		}
-		
-		return condition != null;
+
+		return (condition != null) && canInit;
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/test/runner/EffAssert.java:100]:
```suggestion
		return condition != null && canInit;
```

**Review by AyhamAl-Ali — APPROVED** [review]
Quality changes :zap:

File: src/main/java/ch/njol/skript/test/runner/EffAssert.java (hunk 1 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/test/runner/EffAssert.java:100]: I know logically these are the same, but I like having the () to make it abundantly clear that it's not `condition != (null && canInit)`.

Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/test/runner/EffAssert.java:100]: Ah awwight, fair enough

**Comment by Moderocky** [comment]
I think I've fixed the conflicts on this but please just triple check it looks how you imagine it ought to!

**Comment by sovdeeth** [comment]
> I think I've fixed the conflicts on this but please just triple check it looks how you imagine it ought to!

looks correct

**Pull request merged**